Extend contribution guidelines (new commands, services, verbosity) (#202)
stackitcloud/stackit-cli: CONTRIBUTION.md

## Table of contents

- [Contribute to the STACKIT CLI](#contribute-to-the-stackit-cli)

>   - [Table of contents](#table-of-contents)

Reviewer: Nitpick, but is it common to have the table of contents in the table of contents? :sweat_smile: 
I think we can start with `Developer Guide` at the top

Author: Hm, I'm not sure why but the table of contents is automatically updated every time I save the file, and it includes _all_ headings. I've removed it and saved without formatting

- `make generate-docs`: generate Markdown documentation for every command
- `make test`: run unit tests

### Repository structure

The CLI commands are located under `internal/cmd`, where each folder includes the source code for each subcommand (including their own subcommands). Inside `pkg` you can find several useful packages that are shared by the commands and provide additional functionality such as `flags`, `globalflags`, `tables`, etc.

### Implementing a new command

Let's suppose you want to want to implement a new command `bar`, that would be the direct child of an existing command `stackit foo` (meaning it would be invoked as `stackit foo bar`):

1. You would start by creating a new folder `bar/` inside `internal/cmd/foo/`
2. Following with the creation of a file `bar.go` inside your new folder `internal/cmd/foo/bar/`
   1. The Go package should be similar to the command usage, in this case `package bar` would be an adequate name
   2. Please refer to the [Command file structure](./CONTRIBUTION.md/#command-file-structure) section for details on the strcutre of the file itself

> 3. To register the command `bar` as a child of the exisitng command `foo`, add `cmd.AddCommand(bar.NewCmd(p))` to the `addSubcommands` method of the constructor of the `foo` command

Reviewer: ```suggestion
3. To register the command `bar` as a child of the existing command `foo`, add `cmd.AddCommand(bar.NewCmd(p))` to the `addSubcommands` method of the constructor of the `foo` command
```

>    1. In this case, `p` in this case is the `printer` that is passed from the root command to all subcommands of the tree (refer to the [Outputs, prints and debug logs](./CONTRIBUTION.md/#outputs-prints-and-debug-logs) section for more details regarding the `printer`)

Reviewer: ```suggestion
   1. In this case, `p` is the `printer` that is passed from the root command to all subcommands of the tree (refer to the [Outputs, prints and debug logs](./CONTRIBUTION.md/#outputs-prints-and-debug-logs) section for more details regarding the `printer`)
```

Please remeber to run `make generate-docs` after your changes to keep the commands' documentation updated.

#### Command file structure

Below is a typical structure of a CLI command:

```go
package bar

import (
	(...)
)

// Define consts for command flags
const (
   someArg = "MY_ARG"

> 	someFlag = "my-flag"

Reviewer: wrong indentation?

)

// Struct to model user input (arguments and/or flags)
type inputModel struct {
	*globalflags.GlobalFlagModel
	MyArg string

	MyFlag *string
}

// "bar" command constructor
func NewCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "bar",
		Short: "Short description of the command (is shown in the help of parent command)",
		Long:  "Long description of the command. Can contain some more information about the command usage. It is shown in the help of the current command.",

> 		Args:  args.SingleArg(instanceIdArg, (utils.ValidateUUID)), // Validate argument

Reviewer: ```suggestion
		Args:  args.SingleArg(someArg, utils.ValidateUUID), // Validate argument, with an optional validation function
```

		Example: examples.Build(
			examples.NewExample(
				`Do something with command "bar"`,
				"$ stackit foo bar arg-value --my-flag flag-value"),
			...
		),
		RunE: func(cmd *cobra.Command, args []string) error {
			ctx := context.Background()
			model, err := parseInput(cmd, args)
			if err != nil {
				return err
			}

			// Configure API client
			apiClient, err := client.ConfigureClient(cmd)
			if err != nil {
				return err
			}

			// Call API
			req := buildRequest(ctx, model, apiClient)
         if err != nil {
				return fmt.Errorf("(...): %w", err)
			}
			resp, err := req.Execute()
			if err != nil {
				return fmt.Errorf("(...): %w", err)
			}

         projectLabel, err := projectname.GetProjectName(ctx, cmd)
				if err != nil {
					projectLabel = model.ProjectId
				}

>          // Check API response "resp" an output accordingly

Reviewer: ```suggestion
         // Check API response "resp" and output accordingly
```

         if resp.Item == nil {
            p.Info("(...)", projectLabel)
				return nil
         }
			return outputResult(cmd, model.OutputFormat, instances)
		},
	}

	configureFlags(cmd)
	return cmd
}

// Configure command flags (type, default value, and description)
func configureFlags(cmd *cobra.Command) {
	cmd.Flags().StringP(myFlag, "defaultValue", "My flag description")
}

// Parse user input (arguments and/or flags)
func parseInput(cmd *cobra.Command, inputArgs []string) (*inputModel, error) {
   myArg := inputArgs[0]

	globalFlags := globalflags.Parse(cmd)
	if globalFlags.ProjectId == "" {
		return nil, &errors.ProjectIdError{}
	}

	return &inputModel{
		GlobalFlagModel: globalFlags,
      MyArg            myArg,
		MyFlag:          flags.FlagToStringPointer(cmd, myFlag),
	}, nil
}

// Build request to the API
func buildRequest(ctx context.Context, model *inputModel, apiClient *foo.APIClient) (foo.ApiListInstancesRequest error) {

>    someParam, err := utils.GetParam(model.MyFlag) // Util function to get API parameter based on flag value

Reviewer: this is not the most common case, so for simplicity I would remove it

   if err != nil {
      return fmt.Errorf("(...): %w", err)
   }
	req := apiClient.GetBar(ctx, model.ProjectId, model.MyArg, someParam)
	return req
}

// Output result based on the configured output format
func outputResult(cmd *cobra.Command, outputFormat string, resources []foo.Resource) error {
	switch outputFormat {
	case globalflags.JSONOutputFormat:

> 		details, err := json.MarshalIndent(instances, "", "  ")

Reviewer: ```suggestion
		details, err := json.MarshalIndent(resources, "", "  ")
```

		table.SetHeader("ID", "NAME", "STATE")
		for i := range resources {
			resource := resources[i]
			table.AddRow(*resource.ResourceId, *resource.Name, *resource.State)
		}
		err := table.Display(cmd)
		if err != nil {
			return fmt.Errorf("render table: %w", err)
		}
		return nil
	}
}
```

Please remeber to always add unit tests for `parseInput`, `buildRequest` (in `bar_test.go`), and any other util functions used.

If the new command `bar` is the first command in the CLI using a STACKIT service `foo`, please refer to [Onboarding a new STACKIT service](./CONTRIBUTION.md/#onboarding-a-new-stackit-service).

#### Outputs, prints and debug logs

The CLI has 4 different verbosity levels:

- `error`: For only displaying errors
- `warning`: For displaying user facing warnings _(and all of the above)_
- `info` (default): For displaying user facing info, such as operation success messages and spinners _(and all of the above)_
- `debug`: For displaying structured logs with different levels, including errors _(and all of the above)_

For prints that are specific to a certain log level, you can use the methods defined in the `print` package: `Error`, `Warn`, `Info`, and `Debug`.

> For command outputs that should always be displayed, no matter the defined verbosity, you should use the `print` methods `Outputf` and `Outputln`.

Reviewer: Make it more clear that these methods should only be used for the command output, which are the outputs described by "I run the command to see this"

    2.  Below is an example of a typical `client.go` file structure:

        ```go
        package client

        import (
           (...)
           "github.com/stackitcloud/stackit-sdk-go/services/foo"
        )

        func ConfigureClient(cmd *cobra.Command) (*foo.APIClient, error) {
           var err error
           var apiClient foo.APIClient
           var cfgOptions []sdkConfig.ConfigurationOption

           authCfgOption, err := auth.AuthenticationConfig(cmd, auth.AuthorizeUser)
           if err != nil {
              return nil, &errors.AuthError{}
           }
           cfgOptions = append(cfgOptions, authCfgOption, sdkConfig.WithRegion("eu01")) // Configuring region is needed if "foo" is a regional API

           customEndpoint := viper.GetString(config.fooCustomEndpointKey)

           if customEndpoint != "" {
              cfgOptions = append(cfgOptions, sdkConfig.WithEndpoint(customEndpoint))
           }

           apiClient, err = foo.NewAPIClient(cfgOptions...)
           if err != nil {
              return nil, &errors.AuthError{}
           }

           return apiClient, nil
        }
        ```

> 2.  Add a `FooCustomEndpointKey` key in `internal/pkg/config/config.go` (and add it to `ConfigKeys` and set the to default to `""` using `viper.SetDefault`)

Reviewer: Steps 2 and 3 should come in the beginning. Otherwise in the client we are using consts that are not yet defined

> 2.  Add a `FooCustomEndpointKey` key in `internal/pkg/config/config.go` (and add it to `ConfigKeys` and set the to default to `""` using `viper.SetDefault`)

Reviewer: add a small snippet here maybe?

Author: I think having more snippets might make the document more complex and harder to read, in this case it becomes clear once the user opens the file and sees all other uses for other services

> 3.  Update the `stackit config unset` and `stackit config unset` commands by adding flags to set and unset a custom endpoint for the `foo` service API, respectively, and update their unit tests

Reviewer: also here

Author: Same as above
